File: Data Structures & Algorithms/products-of-array-discluding-self/submission-6.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # [1,2,4,6]
        # [48,24,12,8]

        res = [1] * len(nums)

        prefix=1
        prefixes=[1]*len(nums)
        for i in range(len(nums)):
            res[i] = prefix
            prefix *= nums[i]
        # prefixes=[1,1,2,8]

        suffix = 1
        suffixes=[1]*len(nums)
        for i in range(len(nums)-1,-1,-1):
            res[i] = suffix * res[i]
            suffix *= nums[i]
        
        # suffixes=[48,24,6,1]

        return res

            





        

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-6.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0;
        int r = nums.size() - 1;

        while (l < r) {
            int m = (l + r) / 2;
            if (nums[m] > nums[r]) {
                l = m + 1;
            } else if (nums[m] < nums[r]) {
                r = m;
            }
        }

        int pivot = l;
        l = 0;
        r = nums.size() - 1;

        if (target >= nums[pivot] && target <= nums[nums.size()-1]) {
            l = pivot;
            r = nums.size()-1;
        } else {
            l = 0;
            r = pivot - 1;
        }

        while (l <= r) {
            int m = l + (r-l) / 2;
            if (nums[m] == target) {
                return m;
            } else if (nums[m] < target) {
                l = m+1;
            } else if (nums[m] > target) {
                r = m - 1;
            }
        }

        return -1;
        
    }
};